File: tape/sdk/go/reactors.go
package tape

import (
	"context"
	"encoding/json"
	"io"
	"sync"
	"time"

	pb "github.com/vamsiramakrishnan/durable-agents/tape/sdk/go/tapepb"
)

// RedriveFn — how a reactor re-invokes a stalled run. On Vertex AI Agent
// Engine the implementation calls :streamQuery; on Cloud Run / GKE / locally
// it's typically a wrapper that hands control back to the agent process.
type RedriveFn func(ctx context.Context, run *pb.RunState) error

// ──── registries (compensators / status checks) ─────────────────────────────

var (
	compMu        sync.RWMutex
	compensators  = map[string]func(ctx context.Context, payload []byte) error{}
	statusMu      sync.RWMutex
	statusChecks  = map[string]func(ctx context.Context, key string) (StatusCheckResult, error){}
)

// StatusCheckResult — the per-tool status_check returns one of these. Found =
// "the counterparty acknowledges this idempotency_key (the request did land)".
// ResponseJSON is what to record on the effect when Found.
type StatusCheckResult struct {
	Found        bool
	ResponseJSON string
}

func RegisterCompensator(kind string, fn func(ctx context.Context, payload []byte) error) {
	compMu.Lock()
	defer compMu.Unlock()
	compensators[kind] = fn
}
func GetCompensator(kind string) func(ctx context.Context, payload []byte) error {
	compMu.RLock()
	defer compMu.RUnlock()
	return compensators[kind]
}
func RegisterStatusCheck(tool string, fn func(ctx context.Context, key string) (StatusCheckResult, error)) {
	statusMu.Lock()
	defer statusMu.Unlock()
	statusChecks[tool] = fn
}
func GetStatusCheck(tool string) func(ctx context.Context, key string) (StatusCheckResult, error) {
	statusMu.RLock()
	defer statusMu.RUnlock()
	return statusChecks[tool]
}

// ──── recovery ──────────────────────────────────────────────────────────────

func RecoverOnce(ctx context.Context, c *Client, redrive RedriveFn) ([]string, error) {
	resp, err := c.ListRunsToRecover(ctx, 50)
	if err != nil {
		return nil, err
	}
	out := make([]string, 0, len(resp.Runs))
	for _, r := range resp.Runs {
		if redrive != nil {
			if err := redrive(ctx, r); err != nil {
				continue
			}
		}
		out = append(out, r.RunId)
	}
	return out, nil
}

// ──── reconciler ────────────────────────────────────────────────────────────

func ReconcileOnce(ctx context.Context, c *Client, reconcilePendingAfter time.Duration) (int, error) {
	includePending := reconcilePendingAfter > 0
	olderThan := int64(0)
	if includePending {
		olderThan = time.Now().Add(-reconcilePendingAfter).UnixMilli()
	}
	resp, err := c.ListPendingEffects(ctx, olderThan, includePending, true, 500)
	if err != nil {
		return 0, err
	}
	n := 0
	for _, e := range resp.Effects {
		check := GetStatusCheck(e.ToolName)
		if check == nil {
			continue
		}
		res, err := check(ctx, e.IdempotencyKey)
		if err != nil {
			continue
		}
		if res.Found {
			if _, err := c.ReconcileEffect(ctx, e.RunId, e.IdempotencyKey, EffectStatusConfirmed, res.ResponseJSON, ""); err == nil {
				n++
			}
		} else if int32(e.Status) == EffectStatusUnknown {
			if _, err := c.ReconcileEffect(ctx, e.RunId, e.IdempotencyKey, EffectStatusFailed, "", `{"reconciled":"absent at counterparty"}`); err == nil {
				n++
			}
		}
	}
	return n, nil
}

// ──── timer reactor ─────────────────────────────────────────────────────────

func FireDueTimersOnce(ctx context.Context, c *Client, redrive RedriveFn) (int, error) {
	resp, err := c.ListDueTimers(ctx, 0, 500, true)
	if err != nil {
		return 0, err
	}
	n := 0
	for _, t := range resp.Timers {
		switch t.Kind {
		case "gate_timeout":
			var payload map[string]any
			_ = json.Unmarshal([]byte(t.PayloadJson), &payload)
			gate, _ := payload["gate"].(string)
			res := map[string]any{"timed_out": true}
			if r, ok := payload["resolution"].(map[string]any); ok {
				for k, v := range r {
					res[k] = v
				}
			}
			rj, _ := json.Marshal(res)
			if _, err := c.SendSignal(ctx, SendSignalOpts{RunID: t.RunId, GateName: gate, ResolutionJSON: string(rj)}); err == nil {
				n++
			}
		case "redrive":
			if redrive != nil {
				run, err := c.GetRun(ctx, t.RunId)
				if err == nil {
					_ = redrive(ctx, run)
					n++
				}
			}
		}
	}
	return n, nil
}

// ──── the loop ──────────────────────────────────────────────────────────────

type RunReactorsOptions struct {
	Redrive                RedriveFn
	Recover                bool
	Reconcile              bool
	Timers                 bool
	Interval               time.Duration
	ReconcilePendingAfter  time.Duration
	Once                   bool
	OnTick                 func(tick map[string]any)
}

func RunReactors(ctx context.Context, c *Client, opt RunReactorsOptions) error {
	if opt.Interval == 0 {
		opt.Interval = 2 * time.Second
	}
	if !opt.Recover && !opt.Reconcile && !opt.Timers {
		opt.Recover, opt.Reconcile, opt.Timers = true, true, true
	}
	for {
		tick := map[string]any{}
		if opt.Recover && opt.Redrive != nil {
			if recovered, err := RecoverOnce(ctx, c, opt.Redrive); err != nil {
				tick["recover_error"] = err.Error()
			} else {
				tick["recovered"] = recovered
			}
		}
		if opt.Reconcile {
			if n, err := ReconcileOnce(ctx, c, opt.ReconcilePendingAfter); err != nil {
				tick["reconcile_error"] = err.Error()
			} else {
				tick["reconciled"] = n
			}
		}
		if opt.Timers {
			if n, err := FireDueTimersOnce(ctx, c, opt.Redrive); err != nil {
				tick["timer_error"] = err.Error()
			} else {
				tick["timers_fired"] = n
			}
		}
		if opt.OnTick != nil {
			opt.OnTick(tick)
		}
		if opt.Once {
			return nil
		}
		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-time.After(opt.Interval):
		}
	}
}

// ──── WAL fan-out ───────────────────────────────────────────────────────────

// RunEventFanout is the legacy timestamp-cursored WAL fan-out. Preserved for
// back-compat — new code should use RunEventFanoutBySubject (global_seq
// cursored) or RunEventFanoutWith (the option-struct form supporting both
// cursors plus a subject pattern).
//
// Deprecated: prefer RunEventFanoutBySubject for global_seq cursoring.
func RunEventFanout(ctx context.Context, c *Client, fromTsMs int64, runID, kind string, sink func(entry *pb.EventEntry) error) error {
	stream, err := c.SubscribeEvents(ctx, fromTsMs, runID, kind)
	if err != nil {
		return err
	}
	for {
		entry, err := stream.Recv()
		if err == io.EOF {
			return nil
		}
		if err != nil {
			return err
		}
		if err := sink(entry); err != nil {
			return err
		}
	}
}

// RunEventFanoutBySubject tails the journal via SubscribeBySubject, cursored
// on `global_seq`. The matching subject pattern follows the path-style grammar
// (`*` = one segment, `**` = trailing segments). `predicateCEL` is an optional
// server-side CEL filter on the envelope.
func RunEventFanoutBySubject(ctx context.Context, c *Client, subjectPattern, predicateCEL string, fromGlobalSeq int64, sink func(entry *pb.EventEntry) error) error {
	stream, err := c.SubscribeBySubject(ctx, subjectPattern, predicateCEL, fromGlobalSeq)
	if err != nil {
		return err
	}
	for {
		entry, err := stream.Recv()
		if err == io.EOF {
			return nil
		}
		if err != nil {
			return err
		}
		if err := sink(entry); err != nil {
			return err
		}
	}
}

// RunEventFanoutOpts mirrors SubscribeEventsOpts: pass `FromGlobalSeq` /
// `SubjectPattern` for the event-bus path or `FromTsMs` / `RunID` / `Kind`
// for the legacy filters.
type RunEventFanoutOpts struct {
	FromTsMs       int64
	RunID          string
	Kind           string
	FromGlobalSeq  int64
	SubjectPattern string
}

// RunEventFanoutWith is the option-struct form of RunEventFanout. Honours
// both the legacy and the event-bus filter fields.
func RunEventFanoutWith(ctx context.Context, c *Client, o RunEventFanoutOpts, sink func(entry *pb.EventEntry) error) error {
	stream, err := c.SubscribeEventsWith(ctx, SubscribeEventsOpts{
		FromTsMs:       o.FromTsMs,
		RunID:          o.RunID,
		Kind:           o.Kind,
		FromGlobalSeq:  o.FromGlobalSeq,
		SubjectPattern: o.SubjectPattern,
	})
	if err != nil {
		return err
	}
	for {
		entry, err := stream.Recv()
		if err == io.EOF {
			return nil
		}
		if err != nil {
			return err
		}
		if err := sink(entry); err != nil {
			return err
		}
	}
}

File: tape/sdk/typescript/src/reactors.ts
// Reactors — the WAL-driven side. Each is idempotent (the lease + replay
// properties make a double-run harmless), so run as many copies as you like.

import { TapeClient, EffectStatus, DEFAULT_URL } from './client.ts';
import { getStatusCheck } from './effect.ts';

export type RedriveFn = (run: any) => Promise<void> | void;

// ── recovery ────────────────────────────────────────────────────────────────
export async function recoverOnce(opts: { url?: string; redriveFn: RedriveFn; limit?: number }): Promise<Array<{ runId: string; invocationId: string }>> {
  const c = new TapeClient(opts.url ?? DEFAULT_URL);
  try {
    const resp: any = await c.listRunsToRecover({ limit: opts.limit ?? 50 });
    const out: Array<{ runId: string; invocationId: string }> = [];
    for (const r of resp.runs ?? []) {
      await opts.redriveFn(r);
      out.push({ runId: r.runId, invocationId: r.invocationId });
    }
    return out;
  } finally { c.close(); }
}

// ── reconciler ──────────────────────────────────────────────────────────────
export async function reconcileOnce(opts: { url?: string; reconcilePendingAfterMs?: number } = {}): Promise<Array<{ key: string; resolved: string }>> {
  const c = new TapeClient(opts.url ?? DEFAULT_URL);
  try {
    const includePending = (opts.reconcilePendingAfterMs ?? 0) > 0;
    const olderThanMs = includePending ? Date.now() - (opts.reconcilePendingAfterMs ?? 0) : 0;
    const resp: any = await c.listPendingEffects({ olderThanMs, includePending, includeUnknown: true, limit: 500 });
    const out: Array<{ key: string; resolved: string }> = [];
    for (const e of resp.effects ?? []) {
      const check = getStatusCheck(e.toolName);
      if (!check) continue;
      let res: any;
      try { res = await check(e.idempotencyKey); }
      catch (ex) { out.push({ key: e.idempotencyKey, resolved: `check-error: ${ex}` }); continue; }
      const found = typeof res === 'object' && res !== null ? Boolean(res.found ?? true) : Boolean(res);
      if (found) {
        await c.reconcileEffect({ runId: e.runId, idempotencyKey: e.idempotencyKey,
          resolvedStatus: EffectStatus.CONFIRMED, responseJson: JSON.stringify(res ?? {}) });
        out.push({ key: e.idempotencyKey, resolved: 'confirmed' });
      } else if (e.status === EffectStatus.UNKNOWN) {
        await c.reconcileEffect({ runId: e.runId, idempotencyKey: e.idempotencyKey,
          resolvedStatus: EffectStatus.FAILED, errorJson: JSON.stringify({ reconciled: 'absent at counterparty' }) });
        out.push({ key: e.idempotencyKey, resolved: 'failed' });
      }
    }
    return out;
  } finally { c.close(); }
}

// ── timer reactor ───────────────────────────────────────────────────────────
export async function fireDueTimersOnce(opts: { url?: string; redriveFn?: RedriveFn; onTimer?: (t: any) => Promise<void> | void } = {}): Promise<Array<{ runId: string; timerId: string; kind: string; action: string }>> {
  const c = new TapeClient(opts.url ?? DEFAULT_URL);
  try {
    const resp: any = await c.listDueTimers({ claim: true, limit: 500 });
    const out: Array<{ runId: string; timerId: string; kind: string; action: string }> = [];
    for (const t of resp.timers ?? []) {
      let payload: any = {};
      try { payload = t.payloadJson ? JSON.parse(t.payloadJson) : {}; } catch {}
      let action = 'ignored';
      try {
        if (t.kind === 'gate_timeout') {
          await c.sendSignal({ runId: t.runId, gateName: payload.gate ?? '',
            resolutionJson: JSON.stringify({ timedOut: true, ...(payload.resolution ?? {}) }) });
          action = `signalled ${payload.gate} (timeout)`;
        } else if (t.kind === 'redrive' && opts.redriveFn) {
          const run = await c.getRun(t.runId);
          await opts.redriveFn(run);
          action = 're-driven';
        } else if (opts.onTimer) {
          await opts.onTimer(t); action = 'delegated';
        }
      } catch (e) { action = `error: ${e}`; }
      out.push({ runId: t.runId, timerId: t.timerId, kind: t.kind, action });
    }
    return out;
  } finally { c.close(); }
}

// ── the loop ────────────────────────────────────────────────────────────────
export interface RunReactorsOptions {
  url?: string;
  redriveFn?: RedriveFn;
  recover?: boolean;
  reconcile?: boolean;
  timers?: boolean;
  intervalMs?: number;
  reconcilePendingAfterMs?: number;
  once?: boolean;
  onTick?: (tick: any) => void;
}

export async function runReactors(opts: RunReactorsOptions = {}): Promise<void> {
  const intervalMs = opts.intervalMs ?? 2000;
  const recover = opts.recover ?? true;
  const reconcile = opts.reconcile ?? true;
  const timers = opts.timers ?? true;
  // eslint-disable-next-line no-constant-condition
  while (true) {
    const tick: any = {};
    if (recover && opts.redriveFn) {
      try { tick.recovered = await recoverOnce({ url: opts.url, redriveFn: opts.redriveFn }); }
      catch (e) { tick.recoverError = String(e); }
    }
    if (reconcile) {
      try { tick.reconciled = await reconcileOnce({ url: opts.url, reconcilePendingAfterMs: opts.reconcilePendingAfterMs }); }
      catch (e) { tick.reconcileError = String(e); }
    }
    if (timers) {
      try { tick.timersFired = await fireDueTimersOnce({ url: opts.url, redriveFn: opts.redriveFn }); }
      catch (e) { tick.timerError = String(e); }
    }
    opts.onTick?.(tick);
    if (opts.once) return;
    await new Promise((r) => setTimeout(r, intervalMs));
  }
}

// ── WAL fan-out ─────────────────────────────────────────────────────────────
export async function runEventFanout(opts: { url?: string; sink: (entry: any) => Promise<void> | void; fromTsMs?: number; runId?: string; kind?: string }): Promise<void> {
  const c = new TapeClient(opts.url ?? DEFAULT_URL);
  try {
    for await (const entry of c.subscribeEvents({ fromTsMs: opts.fromTsMs, runId: opts.runId, kind: opts.kind })) {
      await opts.sink(entry);
    }
  } finally { c.close(); }
}
